relay.go

```diff
@@ -694,3 +770,91 @@ func validateRelayMaxTimeout(d time.Duration, logger Logger) time.Duration {
 	).Warn("Configured RelayMaxTimeout is invalid, using default instead.")
 	return _defaultRelayMaxTimeout
 }
+
+type relayFragmentSender struct {
+	callReq            *lazyCallReq
+	framePool          FramePool
+	frameReceiver      frameReceiver
+	failRelayItemFunc  func(items *relayItems, id uint32, failure string)
+	outboundRelayItems *relayItems
+	origID             uint32
+	sentBytes          uint
+	call               RelayCall
+}
+
+var fragmentSenderPool = sync.Pool{New: func() interface{} { return &relayFragmentSender{} }}
+
+func (r *Relayer) newFragmentSender(dstRelay frameReceiver, cr *lazyCallReq, origID uint32, call RelayCall) *relayFragmentSender {
+	fs := fragmentSenderPool.Get().(*relayFragmentSender)
+	fs.callReq = cr
+	fs.framePool = r.conn.opts.FramePool
+	fs.frameReceiver = dstRelay
+	fs.failRelayItemFunc = r.failRelayItem
+	fs.outboundRelayItems = r.outbound
+	fs.origID = origID
+	fs.sentBytes = 0
+	fs.call = call
+	return fs
+}
+
+func (rfs *relayFragmentSender) newFragment(initial bool, checksum Checksum) (*writableFragment, error) {
+	frame := rfs.framePool.Get()
+	frame.Header.ID = rfs.callReq.Header.ID
+	if initial {
+		frame.Header.messageType = messageTypeCallReq
+	} else {
+		frame.Header.messageType = messageTypeCallReqContinue
+	}
+
+	contents := typed.NewWriteBuffer(frame.Payload[:])
+
+	// flags:1
+	flagsRef := contents.DeferByte()
+	flagsRef.Update(rfs.callReq.Payload[_flagsIndex])
+
+	if initial {
+		// Copy all data before the checksum for the initial frame
+		contents.WriteBytes(rfs.callReq.Payload[_flagsIndex+1 : rfs.callReq.checksumTypeOffset])
+	}
+
+	// checksumType:1
+	contents.WriteSingleByte(byte(checksum.TypeCode()))
+
+	// checksum: checksum.Size()
+	checksumRef := contents.DeferBytes(checksum.Size())
+
+	if initial {
+		// arg1~1: write arg1 to the initial frame
+		contents.WriteUint16(uint16(len(rfs.callReq.method)))
+		contents.WriteBytes(rfs.callReq.method)
+		checksum.Add(rfs.callReq.method)
+	}
+
+	wf := writableFragmentPool.Get().(*writableFragment)
+	wf.flagsRef = flagsRef
+	wf.checksumRef = checksumRef
+	wf.checksum = checksum
+	wf.contents = contents
+	wf.frame = frame
+
+	return wf, contents.Err()
+}
+
+func (rfs *relayFragmentSender) flushFragment(wf *writableFragment) error {
+	defer writableFragmentPool.Put(wf)
+
+	wf.frame.Header.SetPayloadSize(uint16(wf.contents.BytesWritten()))
+	rfs.sentBytes += uint(wf.frame.Header.FrameSize())
+
+	sent, failure := rfs.frameReceiver.Receive(wf.frame, requestFrame)
+	if !sent {
+		rfs.failRelayItemFunc(rfs.outboundRelayItems, rfs.origID, failure)
+		return nil
+	}
+	return nil
+}
+
+func (rfs *relayFragmentSender) doneSending() {
+	rfs.call.SentBytes(rfs.sentBytes)
+	fragmentSenderPool.Put(rfs)
```

Comment:
same as above, are we 100% sure that the fragment sender is never used after?

maybe this should be a TODO with a follow up when we're definite, and we may want to put safety checks to ensure it's not reused